File: Node/test-pattern.js
#!/usr/bin/env node

/**
 * @file Script to validate JSON files against given schema
 *
 * Usage: node <script-path>.js json/file/path.json schema/file/path.json
 *
 */

const { exit } = require('process')
const fs = require('fs')

const [, , filePath, schemaPath] = process.argv

const json = JSON.parse(fs.readFileSync(filePath, { encoding: 'utf-8' }))
const schema = JSON.parse(fs.readFileSync(schemaPath, { encoding: 'utf-8' }))

const pattern = schema.properties.test_values.items.properties.value.pattern
const r = new RegExp(pattern)

console.log('Current regex to test:', '\n', pattern)

let failed = false

json.test_values.forEach(element => {
    const result = (r.exec(element.value) != null)
    failed = failed | (result != element.assertion) 
    console.log(result, '\t', element)
});

console.log("\nOverall result", '\t', failed ? "Error occurred" : "Test successful")

if (failed) {
    exit(1)
}

File: Python/test-pattern.py
#!/usr/bin/env python3

import sys
import json
import re

def main():
    if len(sys.argv) < 3:
        print("Usage: python3 <script-path>.py json/file/path.json schema/file/path.json")
        sys.exit(1)

    file_path = sys.argv[1]
    schema_path = sys.argv[2]

    with open(file_path, 'r', encoding='utf-8') as file:
        json_data = json.load(file)

    with open(schema_path, 'r', encoding='utf-8') as schema_file:
        schema_data = json.load(schema_file)

    pattern = schema_data['properties']['test_values']['items']['properties']['value']['pattern']
    regex = re.compile(pattern)

    print("Current regex to test:\n", pattern)

    failed = False

    for element in json_data['test_values']:
        result = bool(regex.match(element['value']))
        failed |= (result != element['assertion'])
        print(result, "\t", element)

    print("\nOverall result:\t", "Error occurred" if failed else "Test successful")

    if failed:
        sys.exit(1)

if __name__ == "__main__":
    main()